splitgpg2/__init__.py

```diff
@@ -0,0 +1,804 @@
+            untrusted_line, untrusted_rest = \
+                self.untrusted_agent_response_buffer.split(b'\n', 1)
+            self.untrusted_agent_response_buffer = untrusted_rest
+            # convert from bytesarray to bytes to be hashable
+            untrusted_line = bytes(untrusted_line)
+            try:
+                self.handle_agent_response(untrusted_line=untrusted_line)
+            except Exception as e:  # pylint: disable=broad-except
+                # propagate exception back to self.send_command
+                self.agent_responded.set_exception(e)
+                break
+
+    def handle_agent_response(self, *, untrusted_line):
+        try:
+            untrusted_res, untrusted_args = untrusted_line.split(b' ', 1)
+        except ValueError:
+            untrusted_res = untrusted_line
+            untrusted_args = None
+        if untrusted_res in (b'D', b'S'):
+            # passthrough to the client
+            self.server.client_write(untrusted_line + b'\n')
+        elif untrusted_res in (b'OK', b'ERR'):
+            # passthrough to the client and signal command complete
+            self.server.client_write(untrusted_line + b'\n')
+            try:
+                self.agent_responded.set_result(None)
+            except asyncio.InvalidStateError:
+                if untrusted_res == b'ERR':
+                    # agent may report multiple errors
+                    pass
+                raise
+        elif untrusted_res == b'INQUIRE':
+            if not untrusted_args:
+                raise Filtered
+            asyncio.ensure_future(self.handle_agent_inquire(
+                untrusted_args=untrusted_args))
+        else:
+            raise ProtocolError('unexpected gpg-agent response')
+
+    async def handle_agent_inquire(self, *, untrusted_args):
+        try:
+            try:
+                untrusted_inq, untrusted_inq_args = untrusted_args.split(b' ', 1)
+            except ValueError:
+                untrusted_inq = untrusted_args
+                untrusted_inq_args = None
+            try:
+                inquire = self.expected_inquires[untrusted_inq]
+            except KeyError:
+                raise Filtered
+            await inquire(untrusted_args=untrusted_inq_args)
+        except Exception as e:  # pylint: disable=broad-except
+            # propagate exception back to self.send_command
+            if not self.agent_responded.done():
+                self.agent_responded.set_exception(e)
+
+    # region INQUIRE commands sent from gpg-agent
+    #
+
+    async def inquire_KEYPARAM(self, untrusted_args):
+        if untrusted_args:
+            self.server.abort('unexpected arguments to KEYPARAM inquire')
+        await self.server.send_inquire(b'KEYPARAM', {
+            b'D': self.inquire_command_D,
+            b'END': self.inquire_command_END,
+        })
+
+    async def inquire_PINENTRY_LAUNCHED(self, untrusted_args):
+        args = untrusted_args
```

Comment:
This needs sanitization. The ruby code checked for one or more decimal digits.

Reply:
This comes from local server and I think isn't controlled by the client.
Specifically, just allowing digits isn't enough. This is what gpg-agent from GnuPG 2.2.13 (fedora 29) responds:
```
INQUIRE PINENTRY_LAUNCHED 4480 gtk2 1.1.0 /dev/pts/2 xterm-256color :0
```

Reply:
Aha. Yeah I never use the GENKEY or PINENTRY code paths, so I don't notice such changes.

> This comes from local server and I think isn't controlled by the client.

Looking at the code this seems to be correct. Plus: We only send it back, so even malicious content shouldn't matter much here.

So I'm fine with the code but please add a comment like: "This comes from the local agent and shouldn't be controlled by the untrusted client. Additionally the only thing we do with it is to send it back to the client."